File: CircularArrayLoop.cpp
#include<iostream>
#include<vector>

using namespace std;

class Solution {
public:
    bool circularArrayLoop(vector<int>& nums) {
    	int n = nums.size();
        for(int i = 0; i < n; i++) {
        	if(nums[i] == 0) continue;
        	int k = i;
        	bool loop1_flag = false;
        	vector<bool> visited(n, false);
        	while(nums[k] != 0) {
        		int next = (k + nums[k]) % n;
        		if(next < 0) next += n;
        		cout<<"k: "<<k<<endl;
        		cout<<"next: "<<next<<endl;
        		if(k == next) loop1_flag = true;
        		if(nums[k] * nums[next] < 0) break;
        		nums[k] = 0;
        		visited[k] = true;
        		k = next;
			}
			if(!loop1_flag && nums[k] == 0 && visited[k]) return true;
		}
		return false;
    }
};

int main() {
	
	int n;
	cin>>n;
	
	vector<int> nums;
	for(int i = 0; i < n; i++) {
		int num;
		cin>>num;
		nums.push_back(num);
	}
	
	Solution *solution = new Solution();
	solution->circularArrayLoop(nums) ? cout<<"true" : cout<<"false";
	
	return 0;
	
}

File: CircularArrayLoop.py
class Solution:
    def circularArrayLoop(self, nums) -> bool:
        n = len(nums)
        for i in range(n):
            if nums[i] == 0:
                continue
            k = i
            next = ((i + nums[i]) % n + n) % n
            is_loop1 = False
            visited = [False] * n
            while nums[k] != 0 and nums[k] * nums[next] >= 0:
                if k == next:
                    is_loop1 = True
                visited[k] = True
                nums[k] = 0
                k = next
                next = ((next + nums[next]) % n + n) % n
            if not is_loop1 and visited[k] and nums[k] == 0:
                return True
        return False


if __name__ == '__main__':
    nums = [int(num) for num in input().split()]
    solution = Solution()
    print(solution.circularArrayLoop(nums))
